crates/wasmprinter/src/operator.rs

```diff
@@ -1286,3 +1341,450 @@ impl<'a> VisitOperator<'a> for PrintOperator<'_, '_, '_, '_> {
+    // of the foldedinst phrase. In the presence of multi-value instructions with more than 1 result,
+    // it may not be possible to hit the params target exactly. This doesn't necessarily mean the
+    // Wasm is invalid, but it can't be represented sensibly in folded form.
+    fn handle_plain(&mut self, plain: String, params: u32, mut results: u32) -> Result<()> {
+        match self.control.last_mut() {
+            Some(stack) => {
+                let mut first_param = stack.folded.len();
+                let mut param_count: u32 = 0;
+                if params > 0 {
+                    for (pos, inst) in stack.folded.iter().enumerate().rev() {
+                        param_count = param_count.saturating_add(inst.results);
+                        if param_count == params {
+                            first_param = pos;
+                            break;
+                        } else if param_count > params {
+                            // unfoldable instruction
+                            results = u32::MAX;
+                            break;
+                        }
+                    }
+                }
+
+                let mut inst = FoldedInstruction {
+                    plain,
+                    folded: stack.folded.drain(first_param..).collect(),
+                    results,
+                    offset: self.operator_state.op_offset,
+                };
+                if let Some(hint) = self.branch_hint.take() {
+                    inst.folded.push(hint);
+                }
+                stack.folded.push(inst);
+            }
+            _ => bail!("instruction without enclosing block"),
+        }
+
+        Ok(())
+    }
+
+    // Recurse through the stack and print each folded instruction.
+    pub fn finalize(&mut self) -> Result<()> {
+        if self.control.len() != 1 {
+            bail!("instruction sequence not closed");
+        }
+        for inst in &self.control.last().unwrap().folded {
+            PrintOperatorFolded::print(&mut self.printer, &mut self.original_separator, &inst)?;
+        }
+        Ok(())
+    }
+
+    // Print a folded instruction to the "real" printer. First print the "plain"
+    // instruction, then recursively print each instruction that was folded in to the
+    // foldedinst phrase.
+    fn print(
+        printer: &mut Printer,
+        sep: &mut OperatorSeparator,
+        inst: &FoldedInstruction,
+    ) -> Result<()> {
+        match sep {
+            OperatorSeparator::Newline => printer.newline(inst.offset)?,
+            OperatorSeparator::None => (),
+            OperatorSeparator::NoneThenSpace => *sep = OperatorSeparator::Space,
+            OperatorSeparator::Space => printer.result.write_str(" ")?,
+        }
+
+        printer.result.write_str("(")?;
+        printer.result.write_str(&inst.plain)?;
+        if inst.folded.is_empty() && inst.plain.contains(";;") {
+            printer.newline(inst.offset)?;
+        }
```

Comment:
Mind leaving a small comment as to what this is handling?

Reply:
Done.

```rust
// Wasm line comment (e.g. label annotation) shouldn't comment out the closing parenthesis
```